db/version_set.cc

```diff
@@ -1905,6 +1913,28 @@ void Version::Get(const ReadOptions& read_options, const LookupKey& k,
       sample_file_read_inc(f->file_metadata);
     }
 
+    // XXX is the table reader buried in here null?
+    // f->file_metadata->fd
+    // XXX this should be calling Unref on a cache entry it finds
+    Cleanable unpinner;
+    std::function<Cleanable*()> table_pinner = [&]() -> Cleanable* {
+      Ref();
+      unpinner.RegisterCleanup(&ROCKSDB_NAMESPACE::Unref,
+                               reinterpret_cast<void*>(this), nullptr);
+      return &unpinner;
+    };
+
+    bool version_has_table_reader =
+        f->file_metadata->fd.table_reader != nullptr;
+
+    if (version_has_table_reader) {
+      // the table_reader is owned by this version (or SuperVersion? I'm
+      // guessing these files can be used by multiple versions, since they are
+      // immutable). We can ensure this lives by holding a ref to this
+      // version.
+      get_context.set_table_pinner(&table_pinner);
```

Comment:
Thank you for taking on the problem. I think you are making a very interesting optimization: prevent memcopy for mmapped, uncompressed data (potentially for both of plain table and block based table), so I appreciate your effort. I'm just curious, why do we need another pinner for this, rather than directly registering cleaner to GetContext.pinnable_val_?

Reply:
Never mind. Ignore my previous comment. I misunderstood the code and figured out why it is done this way. Let me think about it a little bit more whether there is a potentially better way.